### PR TITLE
Improve handling of assertion failures with very long conditions

### PR DESCRIPTION
**Pull request by nnethercote**
It's not perfectly clear what the best behaviour is here, but I think this is an improvement.

r? @matthewjasper
cc @m-ou-se

### CODE REVIEW

File: compiler/rustc_builtin_macros/src/assert.rs [hunk 1]
```diff
@@ -85,7 +85,7 @@ pub fn expand_assert<'cx>(
                 DUMMY_SP,
                 Symbol::intern(&format!(
                     "assertion failed: {}",
-                    pprust::expr_to_string(&cond_expr).escape_debug()
+                    pprust::expr_to_string(&cond_expr)
```
Comment (compiler-errors) [inline compiler/rustc_builtin_macros/src/assert.rs:88]: we could, perhaps, just `.replace("\n", "")` on this str?

Comment (nnethercote) [reply compiler/rustc_builtin_macros/src/assert.rs:88]: Why? To make it fit on a single line? But a multi-line expression with have one or more indents in it, and they'll show up in the middle of the one line string.

Comment (compiler-errors) [reply compiler/rustc_builtin_macros/src/assert.rs:88]: Ah, yeah, I guess we would have to prune those idents too. Nvm 😸

**Comment by nnethercote** [comment]
The test failure was here:
```
2023-10-09T13:16:26.0432231Z failures:
2023-10-09T13:16:26.0434317Z 
2023-10-09T13:16:26.0434846Z ---- [ui] tests/ui/macros/assert-long-condition.rs stdout ----
2023-10-09T13:16:26.0437558Z diff of run.stderr:
2023-10-09T13:16:26.0437800Z 
2023-10-09T13:16:26.0438136Z 1	thread 'main' panicked at $DIR/assert-long-condition.rs:7:5:
2023-10-09T13:16:26.0438534Z 2	assertion failed: 1 + 2 + 3 + 4 + 5 + 6 + 7 + 8 + 9 + 10 + 11 + 12 + 13 + 14 + 15 + 16 + 17 + 18
2023-10-09T13:16:26.0438879Z 3	                                + 19 + 20 + 21 + 22 + 23 + 24 + 25 == 0
2023-10-09T13:16:26.0439302Z -	note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
2023-10-09T13:16:26.0439607Z +	stack backtrace:
2023-10-09T13:16:26.0439833Z +	   0: rust_begin_unwind
2023-10-09T13:16:26.0440415Z +	             at /rustc/95049a4a9b4e0ee378a478958fde603ad441ee0d/library/std/src/panicking.rs:597:5
2023-10-09T13:16:26.0441138Z +	   1: core::panicking::panic_fmt
2023-10-09T13:16:26.0441484Z +	             at /rustc/95049a4a9b4e0ee378a478958fde603ad441ee0d/library/core/src/panicking.rs:72:14
2023-10-09T13:16:26.0441802Z +	   2: core::panicking::panic
2023-10-09T13:16:26.0442130Z +	             at /rustc/95049a4a9b4e0ee378a478958fde603ad441ee0d/library/core/src/panicking.rs:127:5
2023-10-09T13:16:26.0442442Z +	   3: assert_long_condition::main
2023-10-09T13:16:26.0443119Z +	   4: core::ops::function::FnOnce::call_once
2023-10-09T13:16:26.0443495Z +	note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
```
The problem is that in some cases `RUST_BACKTRACE=1` is set, which changes the uninteresting parts of the output, which means the `.run.stderr` file doesn't match.

Other assert tests in `tests/ui/macros/` instead use the `error-pattern` header to check the output of assertions. I didn't use that because it chops off whitespace from the start and end of the patterns, which means I couldn't precisely check for the expected `                                + 19 + 20 + 21 + 22 + 23 + 24 + 25 == 0` output.

In other words, a `.run.stderr` file over-specifies the expected output, while an `error-pattern` header under-specifies the expected output. I need something in between, that matches the `assertion failed` lines exactly, but ignores the backtrace lines.

**Comment by Mark-Simulacrum** [comment]
IIRC, you can set a header like exec-env: RUST_BACKTRACE=0 or so? That'll disable the backtrace independent of runner environment.

**Comment by nnethercote** [comment]
> IIRC, you can set a header like exec-env: RUST_BACKTRACE=0 or so? That'll disable the backtrace independent of runner environment.

Great suggestion, thank you. I have updated with this change.

@bors r=matthewjasper

**Comment by nnethercote** [comment]
Here's another example:
```rust
fn very_long_function_name(x: u32) -> u32 { x }

fn main() {
    assert!(
        {
            let x = 1;
            let y = 2;
            very_long_function_name(x) + very_long_function_name(y) 
        } == 
        {
            let x = 0;
            let y = 1;
            very_long_function_name(x) + very_long_function_name(y)
        }
    );
}
```
Previously it would produce a single line of output on failure:
```text
assertion failed: {\n        let x = 1;\n        let y = 2;\n        very_long_function_name(x) + very_long_function_name(y)\n    } ==\n    {\n        let x = 0;\n        let y = 1;\n        very_long_function_name(x) + very_long_function_name(y)\n    }
```
Now it produces something much better:
```text
thread 'main' panicked at s3.rs:4:5:
assertion failed: {
        let x = 1;
        let y = 2;
        very_long_function_name(x) + very_long_function_name(y)
    } ==
    {
        let x = 0;
        let y = 1;
        very_long_function_name(x) + very_long_function_name(y)
    }
```
